packages/grid/x-data-grid/src/components/GridRow.tsx

```diff
@@ -225,11 +221,7 @@ function GridRow(props: React.HTMLAttributes<HTMLDivElement> & GridRowProps) {
     const column = renderedColumns[i];
     const indexRelativeToAllColumns = firstColumnToRender + i;
 
-    const isLastColumn = indexRelativeToAllColumns === visibleColumns.length - 1;
-    const removeLastBorderRight = isLastColumn && hasScrollX && !hasScrollY;
-    const showRightBorder = !isLastColumn
-      ? rootProps.showCellRightBorder
-      : !removeLastBorderRight && rootProps.disableExtendRowFullWidth;
+    const showRightBorder = rootProps.showCellRightBorder;
```

Comment:
I think part of the previous code we should keep. On systems that have 0px width scrollbar (e.g. macOS or mobile), I can still see the border.

![msedge_WI93JzhmFc](https://user-images.githubusercontent.com/42154031/159775963-b496784f-7ff1-482e-8468-100cdb7cfa08.gif)